plonky2/src/iop/generator.rs

```diff
@@ -106,6 +110,52 @@ pub trait WitnessGenerator<F: Field>: 'static + Send + Sync + Debug {
     /// flag is true, the generator will never be run again, otherwise it will be queued for another
     /// run next time a target in its watch list is populated.
     fn run(&self, witness: &PartitionWitness<F>, out_buffer: &mut GeneratedValues<F>) -> bool;
+
+    fn serialize(&self, dst: &mut Vec<u8>) -> IoResult<()>;
+
+    fn deserialize(src: &mut Buffer) -> IoResult<Self>
+    where
+        Self: Sized;
+}
+
+pub trait AnyWitnessGenerator<F: Field>: WitnessGenerator<F> {
+    fn as_any(&self) -> &dyn Any;
+}
+
+impl<T: WitnessGenerator<F>, F: Field> AnyWitnessGenerator<F> for T {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+/// A wrapper around an `Box<AnyWitnessGenerator>`.
+pub struct WitnessGeneratorRef<F: Field>(pub Box<dyn AnyWitnessGenerator<F>>);
+
+impl<F: Field> WitnessGeneratorRef<F> {
+    pub fn new<G: AnyWitnessGenerator<F>>(generator: G) -> WitnessGeneratorRef<F> {
+        WitnessGeneratorRef(Box::new(generator))
+    }
+}
+
+impl<F: Field> PartialEq for WitnessGeneratorRef<F> {
+    fn eq(&self, other: &Self) -> bool {
+        let mut buf1 = Vec::new();
+        let mut buf2 = Vec::new();
+        self.0.serialize(&mut buf1).unwrap();
+        other.0.serialize(&mut buf2).unwrap();
+
+        buf1 == buf2
+    }
+}
```

Comment:
Definitely not a fan of this. We could alternatively just check that both "id" match, but that would only enforce that we use the same ordering of generators. Maybe that would be acceptable?

Reply:
Given that checking equality by "id" was already done for gates, I opted for this as well here.